### PR TITLE
#12, #78: Extend command API

### PR DESCRIPTION
**Pull request by martin-fleck-at**
- Make client commands more extensible
-- Use String for type argument instead of enum
-- Add generic properties to provide custom information

- Make command codec injectable with client contributions
-- Clients can contribute translation for their commands based on type
-- Move previous implementation to custom EMF command contributions
-- Move shared code into dedicated utility class
-- Provide EMFCommandCodec for easier non-DI usability

- Use transactional editing domain and command stack
-- Perform model update on command stack (is now undoable)
-- Add EMF Transaction to target platform and maven build

- Extend Command model with CommandExecutionResult
-- Provide original command and affected EObject in result
-- Provide change recordings in result
-- Allow for custom type and details map in result

- Remove 'edit' command with EMF commands from Java Client API

- Add custom command to Coffee example
-- Command simply adds the given text to all tasks

- Simply discard dangling references on XMI serialization

Client adaptation in https://github.com/eclipse-emfcloud/emfcloud-modelserver-theia/pull/35

### CODE REVIEW

**Review by ndoschek — CHANGES_REQUESTED** [review]
Thank you very much for this excellent contribution Martin!
Overall the functionality seems to work great, and I am a huge fan of the re-worked "command concept"! :+1: 
I am currently still in the process of testing, but meanwhile I found some superficial issues, which might be fixed rather quickly I assume.

File: bundles/org.eclipse.emfcloud.modelserver.edit/src/org/eclipse/emfcloud/modelserver/edit/EMFCommandCodec.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+         RemoveCommand undoRemoveCommand = (RemoveCommand) command;
+         return AddCommand.create(undoRemoveCommand.getDomain(), undoRemoveCommand.getOwner(),
+            undoRemoveCommand.getFeature(), undoRemoveCommand.getResult(), undoRemoveCommand.getIndices()[0]);
+      } else if (command instanceof SetCommand) {
+         SetCommand undoSetCommand = (SetCommand) command;
+         return SetCommand.create(undoSetCommand.getDomain(), undoSetCommand.getOwner(),
+            undoSetCommand.getFeature(), undoSetCommand.getOldValue(), undoSetCommand.getIndex());
+      }
+      return command;
+   }
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply bundles/org.eclipse.emfcloud.modelserver.edit/src/org/eclipse/emfcloud/modelserver/edit/EMFCommandCodec.java:176]: We don't use it anymore, no. I just kept it for people who might still want to use it. But maybe it is better to remove it, you are right.

File: bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultModelController.java [hunk 2]
```diff
@@ -216,49 +215,33 @@ public void getValidationConstraints(final Context ctx, final String modeluri) {
...
-         if (undoSuccess) {
-            success(ctx, "Successful undo.");
-            sessionController.commandExecuted(modeluri, encodings);
-            return;
-         }
+      Optional<CCommandExecutionResult> undoExecution = modelRepository.undo(modeluri);
+      if (undoExecution.isPresent()) {
+         success(ctx, "Successful undo.");
+         Map<String, JsonNode> encodings = getCommandEncodings(modeluri, undoExecution.get());
+         sessionController.commandExecuted(modeluri, encodings);
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultModelController.java:222]: Oh yeah, sure, good idea!

File: bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelRepository.java [hunk 3]
```diff
@@ -60,17 +61,13 @@
...
+   Optional<CCommandExecutionResult> undo(String modeluri);
 
    CCommand getRedoCommand(String modeluri);
 
-   boolean redo(String modeluri);
-
-   void addTemporaryCommandResource(String modeluri, Resource resource, CCommand command);
-
-   void removeTemporaryCommandResource(String modeluri, Resource resource);
+   Optional<CCommandExecutionResult> redo(String modeluri);
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelRepository.java]: I cleaned those methods up, thank you!

File: bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelServerEditingDomain.java [hunk 4]
```diff
@@ -56,11 +55,9 @@ public Command getUndoCommand() {
        * As we manage the command stack locally, and our clients will most likely not make use of any concept similar to
        * a command stack, it is necessary to provide an update which holds the command to undo the previous changes as
        * an incrementalUpdate.
-       * In the case of an Undo we need to invert this command, which is currently based on the already implemented
-       * commands in the DefaultCommandCodec.
        */
       if (canUndo()) {
-         return createInverseCommand(commandStack.getUndoCommand());
+         return commandStack.getUndoCommand();
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelServerEditingDomain.java]: I modified the code ab it and cleaned it up, thank you!

**Comment by martin-fleck-at** [comment]
@ndoschek Thank you very much for your review, I took another look at the things you mentioned and pushed a new follow-up commit that fixes several things. I also fixed the test cases, at least for me locally they run through :-)

**Review by ndoschek — CHANGES_REQUESTED** [review]
Thank you again for updating the PR and especially for fixing the tests! :+1: 
So far, the functionality looks great to me and it is quite intuitive to adapt to the new concept!
But I noticed a few minor things:
- The `XmiToEObjectSubsubscriptionListener` needs to be updated as well as it still excpects a `CCommand` instead of the new result type.
- Please update the README
- It would be great if we provided a custom command example in Java somewhere, either in the README and/or in the `ExampleModelServerSubscriptionClient` - what do you think?

File: bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultSessionController.java [hunk 5]
```diff
@@ -124,13 +125,13 @@ public void modelUpdated(final String modeluri) {
    }
 
    @Override
-   public void commandExecuted(final String modeluri, final Map<String, JsonNode> encodings) {
+   public void commandExecuted(final String modeluri, final CCommandExecutionResult exeuction) {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultSessionController.java]: Done

File: bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultSessionController.java [hunk 6]
```diff
@@ -124,13 +125,13 @@ public void modelUpdated(final String modeluri) {
...
       Optional<EObject> root = modelRepository.getModel(modeluri);
       if (root.isEmpty()) {
          broadcastError(modeluri, "Could not load changed object");
          return;
       }
-      broadcastIncrementalUpdates(modeluri, encodings);
+      broadcastIncrementalUpdates(modeluri, exeuction);
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultSessionController.java]: Done

**Comment by martin-fleck-at** [comment]
@ndoschek I pushed an update with all the requested changes. I also moved the code a little bit around and renamed CommandCodecContribution to CommandContribution to avoid any codec confusion ;-)

**Review by ndoschek — APPROVED** [review]
Thank you very much Martin for updating the PR, everything works now as expected! :+1: 
My previous confusion about the AddCommand is resolved as well - I tested it again with pure EMF commands and it works great as well!
Thank you again for this great improvement!

**Pull request merged**